ksql-common/src/main/java/io/confluent/ksql/schema/ksql/LogicalSchema.java

```diff
@@ -333,64 +285,75 @@ public String toString() {
+        .filter(withNamespace(Namespace.META).negate())
+        .map(c -> c.toString(formatOptions))
+        .collect(Collectors.joining(", ", "[", "]"));
+  }
 
-    final String values = value.stream()
-        .map(f -> f.toString(formatOptions))
-        .collect(Collectors.joining(", "));
+  private Optional<NamespacedColumn> findNamespacedColumn(
+      final Predicate<NamespacedColumn> predicate
+  ) {
+    return columns.stream()
+        .filter(predicate)
+        .findFirst();
+  }
 
-    final String join = keys.isEmpty() || values.isEmpty()
-        ? ""
-        : ", ";
+  private <T> Map<Namespace, List<T>> byNamespace(final Function<NamespacedColumn, T> mapper) {
+    final Map<Namespace, List<T>> byNamespace = columns.stream()
+        .collect(Collectors.groupingBy(
+            NamespacedColumn::namespace,
+            Collectors.mapping(mapper, Collectors.toList())
+        ));
 
-    return "[" + keys + join + values + "]";
-  }
+    Arrays.stream(Namespace.values())
+        .forEach(ns -> byNamespace.putIfAbsent(ns, ImmutableList.of()));
 
-  private Set<ColumnName> metaColumnNames() {
-    return columnNames(metadata);
+    return byNamespace;
   }
 
-  private Set<ColumnName> keyColumnNames() {
-    return columnNames(key);
-  }
+  private LogicalSchema rebuild(final boolean withMetaAndKeyColsInValue) {
+    final Map<Namespace, List<NamespacedColumn>> byNamespace = byNamespace(Function.identity());
 
-  private Set<ColumnName> metaAndKeyColumnNames() {
-    final Set<ColumnName> names = metaColumnNames();
-    names.addAll(keyColumnNames());
-    return names;
-  }
+    final List<NamespacedColumn> metadata = byNamespace.get(Namespace.META);
+    final List<NamespacedColumn> key = byNamespace.get(Namespace.KEY);
+    final List<NamespacedColumn> value = byNamespace.get(Namespace.VALUE);
 
-  private static Set<ColumnName> columnNames(final List<Column> struct) {
-    return struct.stream()
-        .map(Column::name)
-        .collect(Collectors.toSet());
-  }
+    final ImmutableList.Builder<NamespacedColumn> builder = ImmutableList.builder();
 
-  private static List<Column> addAlias(final SourceName alias, final List<Column> columns) {
-    final ImmutableList.Builder<Column> builder = ImmutableList.builder();
+    builder.addAll(metadata);
+    builder.addAll(key);
+
+    if (withMetaAndKeyColsInValue) {
+      metadata.stream()
+          .map(c -> NamespacedColumn.of(c.column(), Namespace.VALUE))
+          .forEach(builder::add);
 
-    for (final Column col : columns) {
-      builder.add(col.withSource(alias));
+      key.stream()
+          .map(c -> NamespacedColumn.of(c.column(), Namespace.VALUE))
+          .forEach(builder::add);
     }
-    return builder.build();
+
+    value.stream()
+        .filter(c -> !findNamespacedColumn(
+            withName(c.column().name())
+                .and(withNamespace(Namespace.META)
+                    .or(withNamespace(Namespace.KEY)))).isPresent())
```

Comment:
🐞  I think there's a bug here! this is `(name AND meta) OR key` not `name AND (meta OR key)`

Reply:
I think you're mistaken. This is `name AND (meta OR key)`.

Test's would be failing otherwise.

Reply:
I've re-ordered / re-structured so that it's hopefully easier to grok.